lib/swaps/Swaps.ts

```diff
@@ -900,6 +900,16 @@ class Swaps extends EventEmitter {
 
     this.logger.debug(`handleResolveRequest starting with hash ${rHash}`);
 
+    // first check if we have recovered this deal from a previous swap attempt
+    const recoveredSwap = this.swapRecovery.recoveredPreimageSwaps.get(rHash);
+    if (recoveredSwap && recoveredSwap.rPreimage) {
+      recoveredSwap.state = SwapState.Recovered;
+      recoveredSwap.save().catch(this.logger.error);
+      this.swapRecovery.recoveredPreimageSwaps.delete(rHash);
```

Comment:
I'm wondering if this is the correct place to delete swap from memory and mark it as recovered? Should we wait for confirmation that the recovery was successful?

What happens when:
1. swap client requests the preimage
2. xud sends the preimage and deletes it from memory
3. swap client crashes before it manages to claim the payment
4. swap client comes back online and requests the preimage again
5. xud no longer has the preimage in memory (we still have the preimage in the database, but the swap state is marked as recovered)

Reply:
That's a valid concern, it ventures into how xud deals with swap clients crashing/malfunctioning. Technically that can already happen, as we treat a swap deal as "completed" once we send the preimage to raiden and then remove it from memory (we do persist to the database though). It doesn't necessarily have to be a swap that xud has recovered upon startup.

I think it's slightly out of scope to address in this PR, but we could make it so that on all resolve requests if we don't recognize the hash we check the database for completed swaps and return the preimage if we find a match.

Reply:
> I think it's slightly out of scope to address in this PR, but we could make it so that on all resolve requests if we don't recognize the hash we check the database for completed swaps and return the preimage if we find a match.

That works.

Reply:
Created an issue: https://github.com/ExchangeUnion/xud/issues/1264